File: 06 Paradigmas/1166 Torre de Hanoi Novamente/1166.py
def adcionar_bola(haste, bola):
    for vareta in haste:
        try:
            raiz = (vareta[-1] + bola)**(1/2)
            if raiz == int(raiz):
                vareta.append(bola)
                return True
        except:
            vareta.append(bola)
            return True
    return False


cases = int(input())

for i in range(0, cases):

    haste = []
    num = int(input())

    for index in range(0, num):
        haste.append([])

    bola = 1
    numero_bolas = 0
    while True:
        if not adcionar_bola(haste, bola):
            print(numero_bolas)
            break

        numero_bolas += 1
        bola += 1

# https://www.beecrowd.com.br/judge/pt/problems/view/1166
# Ranking 00099º, Time 0.160, Submiss Date 06/03/2023 19:03:55 (Python 3.9)

File: 06 Paradigmas/1166 Torre de Hanoi Novamente/1166.cpp
#include <bits/stdc++.h>

using namespace std;

bool adicionar_bola(vector<vector<int>>& haste, int bola) {
    for (auto& vareta : haste) {
        if (!vareta.empty()) {
            double raiz = sqrt(vareta.back() + bola);
            if (raiz == static_cast<int>(raiz)) {
                vareta.push_back(bola);
                return true;
            }
        } else {
            vareta.push_back(bola);
            return true;
        }
    }
    return false;
}

int main() {
    int cases;
    cin >> cases;

    for (int i = 0; i < cases; ++i) {
        vector<vector<int>> haste;
        int num;
        cin >> num;

        haste.resize(num);

        int bola = 1;
        int numero_bolas = 0;
        while (true) {
            if (!adicionar_bola(haste, bola)) {
                cout << numero_bolas << endl;
                break;
            }
            ++numero_bolas;
            ++bola;
        }
    }

    return 0;
}
